org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/ParameterizedTypeBinding.java

```diff
@@ -1108,26 +1110,71 @@ public boolean isRawSubstitution() {
 
 	@Override
 	public ReferenceBinding[] permittedTypes() {
-		ReferenceBinding[] permTypes = this.type.permittedTypes();
-		List<ReferenceBinding> applicablePermTypes = new ArrayList<>();
-		for (ReferenceBinding pt : permTypes) {
-			ReferenceBinding permittedTypeAvatar = pt;
-			if (pt.isRawType()) {
-				ReferenceBinding ptRef = pt.actualType();
-				ReferenceBinding enclosingType1 = ptRef.enclosingType();
-				if (enclosingType1 != null) {
-					// don't use TypeSystem.getParameterizedType below as this is just for temporary check.
-					ParameterizedTypeBinding ptb = new ParameterizedTypeBinding(ptRef, this.arguments, ptRef.enclosingType(), this.environment);
-					ptb.superclass();
-					ptb.superInterfaces();
-					permittedTypeAvatar = ptb;
+		List<ReferenceBinding> permittedTypes = new ArrayList<>();
+NextPermittedType:
+		for (ReferenceBinding pt : this.type.permittedTypes()) {
+			// Step 1: Gather all type variables that would need to be solved.
+			Map<TypeVariableBinding, TypeBinding> map = new HashMap<>();
+			TypeBinding current = pt;
+			do {
+				if (current.kind() == Binding.GENERIC_TYPE) {
+					for (TypeVariableBinding tvb : current.typeVariables()) {
+						map.put(tvb,  null);
+					}
 				}
-			}
-			if (permittedTypeAvatar.isCompatibleWith(this))
-				applicablePermTypes.add(pt);
+				current = current.enclosingType();
+			} while (current != null);
+
+			// Step 2: Collect substitutes
+			current = this;
+			TypeBinding sooper = pt.findSuperTypeOriginatingFrom(this);
+			do {
+				if (sooper.isParameterizedType()) {
+					if (current.isParameterizedType()) {
+						for (int i = 0, length = sooper.typeArguments().length; i < length; i++) {
+							TypeBinding t = sooper.typeArguments()[i];
+							if (t instanceof TypeVariableBinding tvb) {
```

Comment:
> This would also capture capture bindings (pun intended), but super types aren't captured, right? So what would happen when a wildcard is used? Should you need to call `isTypeArgumentContainedBy()` instead of testing equality?

I don't know. Let me know if you are able to construct a test case - we could address any issues in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3038